accepted/2022/Multiple-Equivalent-Framework-Support-TFM-As-Aliases.md

```diff
@@ -0,0 +1,397 @@
+: error : Invalid restore input. Duplicate frameworks found: 'net472, net472'. Input files: C:\Code\Temp\Aliases\Aliases.csproj. [C:\Code\Temp\Aliases\Aliases.csproj]
+```
+
+Unfortunately this error does not contain a NU code, which would make it easier for customers to find a documentation page explaining the issue and probable fixes.
+Instead, customers who are not aware of this tooling version issue will likely need to search the error message and we have to hope that search engines will direct them to a page that informs them to upgrade to a newer version of NuGet.exe, or ideally switch to `MSBuild -t:restore` or `dotnet restore`.
+
+This problem is not technically unique to this design change, but since assets file changes are uncommon, this feature is significantly more likely to expose this issue than other features.
+
+### Technical explanation
+
+There are many changes required to implement this feature.
+
+1. [Restore output (assets file) changes](#restore-output-assets-file-changes)
+1. [Components iterating target frameworks](#components-iterating-target-frameworks)
+1. [Lock file changes](#lock-file-changes)
+1. [ProjectReference changes](#project-to-project-references)
+1. [Pack changes](#pack-changes)
+1. [Visual Studio Challenges](#visual-studio-challenges)
+
+The changes listed are largely what NuGet needs to implement in code we own, but there will be impacts to other components as well.
+In particular the .NET SDK will need some (hopefully small) changes to be able to use the assets file correctly.
+However, it's likely that other tools, like various Visual Studio components, might have assumptions about TFM uniqueness in the project, which this feature changes.
+Therefore, while the NuGet team works very closely with the .NET SDK team and we will ensure that building projects work, there will increased risk of other features and tools breaking, which would normally not be affected by NuGet changes.
+
+### Restore output (assets file) changes
+
+In PackageReference, NuGet has a contract with the .NET SDK where NuGet writes the assets file and the .NET SDK consumes it.
+
+This feature will require the following changes to the asset file:
+
+1. [Increment Version Number](#assets-file-version)
+1. [Target Framework Pivots](#target-framework-pivots)
+1. [Add List Lengths](#add-list-lengths)
+
+However, PackageReference is also supported by non-SDK style projects, which use [dotnet/NuGet.BuildTools](https://github.com/NuGet.BuildTools) which, despite the name, is owned by the non-SDK project system team, not NuGet.
+Their implementation parses the JSON file directly, rather than using NuGet.ProjectModel.
+Therefore, if we can implement the assets file changes without breaking changes to the NuGet.ProjectModel APIs, the .NET SDK may "just work" when the newer version of NuGet is inserted into the .NET SDK.
+However, the non-SDK project system will fail.
```

Comment:
If we say that it's a non-goal to support TargetFramework aliases from non SDK-style projects (and hence from NuGet.BuildTools), does that simplify things?  NuGet.BuildTools might generate an error if it encounters a newer assets file format without any changes to it, so we might not need as much synchronization.